src/topology/uniform_space/compact_separated.lean

```diff
@@ -0,0 +1,216 @@
+Authors: Patrick Massot
+-/
+import topology.uniform_space.separation
+/-!
+# Compact separated uniform spaces
+
+## Main statements
+
+* `compact_space_uniformity`: On a separated compact uniform space, the topology determines the
+  uniform structure, entourages are exactly the neighborhoods of the diagonal.
+* `uniform_space_of_compact_t2`: every compact T2 topological structure is induced by a uniform
+  structure. This uniform structure is described in the previous item.
+* Heine-Cantor theorem: continuous functions on compact separated uniform spaces with values in
+  uniform spaces are automatically uniformly continuous. There are several variations, the main one
+  is `compact_space.uniform_continuous_of_continuous`.
+
+## Implementation notes
+
+The construction `uniform_space_of_compact_t2` is not declared as an instance, as it would badly
+loop.
+
+## tags
+
+uniform space, uniform continuity, compact space
+-/
+
+open_locale uniformity topological_space filter
+open filter uniform_space set
+
+variables {α β : Type*} [uniform_space α] [uniform_space β]
+
+
+/-!
+### Uniformity on compact separated spaces
+-/
+
+
+lemma compact_space_uniformity [compact_space α] [separated_space α] : 𝓤 α = ⨆ x : α, 𝓝 (x, x) :=
+begin
+  classical,
+  symmetry, refine le_antisymm nhds_le_uniformity _,
+  by_contra H,
+  obtain ⟨V, hV, h⟩ : ∃ V : set (α × α), (∀ x : α, V ∈ 𝓝 (x, x)) ∧ 𝓤 α ⊓ 𝓟 (-V) ≠ ⊥,
+  { rw le_iff_forall_inf_principal_compl at H,
+    push_neg at H,
+    simpa only [mem_supr_sets] using H },
+  let F := 𝓤 α ⊓ 𝓟 (-V),
+  obtain ⟨⟨x, y⟩, hx⟩ : ∃ (x : α × α), cluster_pt x F :=
+    cluster_point_of_compact h,
+  have : cluster_pt (x, y) (𝓤 α) :=
+    hx.of_inf_left,
+  have hxy : x = y := eq_of_uniformity_inf_nhds this,
+  subst hxy,
+  have : cluster_pt (x, x) (𝓟 (-V)) :=
+   hx.of_inf_right,
+  have : (x, x) ∉ interior V,
+  { have : (x, x) ∈ closure (-V), by rwa mem_closure_iff_cluster_pt,
+    rwa closure_compl at this },
+  have : (x, x) ∈ interior V,
+  { rw mem_interior_iff_mem_nhds,
+    exact hV x },
+  contradiction
+end
+
+lemma unique_uniformity_of_compact_t2 {α : Type*} [t : topological_space α] [compact_space α]
+[t2_space α] {u u' : uniform_space α}
+(h : u.to_topological_space = t) (h' : u'.to_topological_space = t) : u = u' :=
+begin
+  apply uniform_space_eq,
+  change uniformity _ = uniformity _,
+  haveI : @compact_space α u.to_topological_space := by rw h ; assumption,
+  haveI : @compact_space α u'.to_topological_space := by rw h' ; assumption,
+  haveI : @separated_space α u := by rwa [separated_iff_t2, h],
+  haveI : @separated_space α u' :=  by rwa [separated_iff_t2, h'],
+  rw compact_space_uniformity,
+  rw compact_space_uniformity,
+  rw [h, h'],
+end
+
+/-- The unique uniform structure inducing a given compact Hausdorff topological structure. -/
+def uniform_space_of_compact_t2 {α : Type*} [topological_space α] [compact_space α] [t2_space α] :
```

Comment:
This is a pretty long declaration, I think that if you split off the longer proofs, over all the readability will be improved.

Reply:
I'm not convinced. Each field of this declaration makes sense only in the context of this declaration. Actually I think this declaration is mostly useless. I added it for completeness at the last minute, because I was writing in the module docstring that this construction was not there, and I feared it would be harder to do it once I wouldn't be in the uniform space mood anymore. This is why the previous PR was a mess: I added a lot of preparatory lemmas only for this declaration.